### PR TITLE
Rowspan support for tables

### LINKED ISSUE #20092: We don't properly support rowspan
Creating tables that have elements with rowspan will correctly place cells in the next row,  but the element will not occupy the extra row.

We need to correctly calculate the element's size (and correctly use that when calculating the row's size)

### PR DESCRIPTION
**Pull request by Manishearth**
fixes #20092

This just contains the first steps.

We apply a naive algorithm: Spanning cells apply a pressure equal to `block_size / rowspan` on each row they are in. We move table row block size computation into the tables, and make it two pass. In the first pass we compute the sizes of each row, and in the
second pass we assign them, adding them up for any involved cells.

This is missing:

 - [x] Accounting for border sizes
 - [x] Applying pressure to rows that are not the row containing the cell
 - [ ] Reducing pressure on future rows if the current row is able to accomodate more of the cell
 - [x] For tables containing both rows and rowgroups, reset the rowspan info when we hit a rowgroup
 - [x] Correctly handle overflowing rowspans

cc @mbrubeck @pcwalton

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/20128)

### CODE REVIEW

**Comment by jdm** [comment]
```
./components/layout/table_row.rs:143: Line is longer than 120 characters
```

**Comment by jdm** [comment]
```
error: unused import: `style::servo::restyle_damage::ServoRestyleDamage`
  --> components/layout/table_row.rs:30:5
   |
30 | use style::servo::restyle_damage::ServoRestyleDamage;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `-D unused-imports` implied by `-D warnings`
error: aborting due to previous error
```

**Comment by Manishearth** [comment]
This works now. It's not _optimal_, both with some inefficiencies during the computation, and some less-than-idealness in the way we distribute rowspan pressure. But it looks pretty good:

image

**Comment by Manishearth** [comment]
And now it's free of known bugs. Could be landed as-is (after we see the test results), though I'd like to see what your thoughts on improving and optimizing this would be.

**Comment by Manishearth** [comment]
I currently can't come up with an efficient way to distribute row pressures that is _balanced_ -- for example it's possible make `incoming_rowspan_data` contain _leftover_ incoming sizes (which each row divides appropriately),  and rejigger the pressure in a second loop over `incoming_rowspan_data`, but this means that earlier rows have more influence on the spanning cell's layout. So if we have a later row that has large cells, it cannot relieve the pressure from earlier rows.

(The general problem seems computationally hard to solve -- probably requiring an SMT solver -- but I'm hoping to come up with an approximation that's good enough and also efficient)

It's also unfortunate that this requires so many loops. A couple things that could be improved are:

 - Merging block size assignment and position calculation passes
 - Having a fast path for non-rowspan where the first two loops are merged. Row size computation is only affected by downwind border sizes when there's rowspan involved
 - We can clamp row_span to the actual allowed span sometime during table construction. I can't see a cheap way to do this and it's not worth it if it's expensive
 - We actually already have all the border size info, on `collapsed_block_direction_border_widths_for_table`. However, this includes the border info for rows within rowgroups, so the indices are all wrong. It would be possible to compute both.

**Comment by Manishearth** [comment]
Removing the WIP since it's technically ready. Unsure if we should land it as is, but it's ready for review either way.

r? @mbrubeck

**Comment by emilio** [comment]
Lol @ 5-pass layout :)

Thanks for working on this stuff Manish!

**Comment by Manishearth** [comment]
I had a plan for doing this super perfectly but it involved like 8 passes :P

This actually only introduces one pass and complicates one more. Previously row layout would do two cell passes. Now we do two row+cell passes in table layout (which is is slightly less efficient). The only new pass is the border sizing one, and that's a cheap row pass. Once I merge the assignment passes this will not have introduced new passes.

**Comment by pcwalton** [comment]
I'm surprised this doesn't cause some tests to pass.

**Comment by Manishearth** [comment]
We haven't run tests yet. I don't think travis does WPT

File: components/layout/table_row.rs [hunk 1]
```diff
@@ -107,96 +106,158 @@ impl TableRowFlow {
...
-            self.block_flow.base.position.size.block = block_size;
+            let row_span;
+            let column_span;
+            let cell_total;
+            {
+                let cell = kid.as_mut_table_cell();
+                row_span = cell.row_span;
+                column_span = cell.column_span as usize;
+                cell_total = cell.total_block_size();
+            }
```
Comment (pcwalton) [inline components/layout/table_row.rs]: My kingdom for NLL…

File: components/layout/table_row.rs [hunk 2]
```diff
@@ -107,96 +106,158 @@ impl TableRowFlow {
...
+        for kid in self.block_flow.base.child_iter_mut() {
+            include_sizes_from_previous_rows(&mut col, &self.incoming_rowspan,
+                                             incoming_rowspan_data, &mut max_block_size,
+                                             &mut largest_leftover_incoming_size);
+            kid.place_float_if_applicable();
+            if !kid.base().flags.is_float() {
+                kid.assign_block_size_for_inorder_child_if_necessary(layout_context,
+                                                                     thread_id,
+                                                                     content_box);
+            }
```
Comment (pcwalton) [inline components/layout/table_row.rs]: While you're here, can you change this `kid.base().flags.is_float()` condition to a debug assert? Children of table rows should never be floating.

**Comment by Manishearth** [comment]
The failure is tests/wpt/mozilla/tests/css/table_rowspan_rowgroup_a.html`, which tests `rowspan=0` ("maximum possible rowspan")

I wasn't aware such a thing existed, and the only reason it passed before was because we didn't handle rowspan so the ref didn't work either and it passed.

I could add handling for rowspan=0 to my code, but I suspect we don't handle it in other places either.

Really it would be nice to make `cell.row_span` equal to the final rowspan -- clipped for overlong rowspans, and extended for short ones.

**Comment by Manishearth** [comment]
@pcwalton any idea of a good place where we can fixup `cell.row_span` without introducing a new loop? We need to know the size of the row group beforehand here.

**Comment by Manishearth** [comment]
Huh, this is the only failure in the tests.

This is surprising. Either we don't have enough rowspan tests, or we're not doing it perfectly.



It's also possible that all current rowspan tests test rowspan against rowspan. Testing rowspans seems like a hard problem.

**Comment by Manishearth** [comment]
Okay, this is only supported in Firefox, and only when _not_ in quirks mode

https://bugzilla.mozilla.org/show_bug.cgi?id=9879

**Comment by pcwalton** [comment]
Can we just remove it from Firefox? (I'm serious.)

**Comment by Manishearth** [comment]
It's specced: https://html.spec.whatwg.org/multipage/tables.html#algorithm-for-processing-rows (step 10), https://www.w3.org/TR/html401/struct/tables.html#adef-rowspan , and probably some more places.

**Comment by Manishearth** [comment]
Looks like Chrome recently implemented it :|

https://bugs.chromium.org/p/chromium/issues/detail?id=180722

**Comment by Manishearth** [comment]
Yeah, it works in Canary. Damn.

**Comment by Manishearth** [comment]
I'm deferring rowspan=0 support, I talked with dbaron and posted an intent to unship on dev-platform. If that goes through we don't need to support it.

There don't seem to be tests because testing this is hard due to divergence in the exact row sizing algorithms. We could write some testharness tests that ensure that the heights match up (and test for other similar variants), however.

**Comment by pcwalton** [comment]
@Manishearth You could also write a != reftest that tests that rowspan does *something* (i.e. test that a table cell without rowspan is different from a table cell with rowspan).

**Comment by Manishearth** [comment]
Oh, good idea.

**Comment by mbrubeck** [comment]
r=mbrubeck with test expectations updated, and some sort of minimal test added

**Comment by Manishearth** [comment]
@bors-servo r=mbrubeck

It seems like we're not unshipping rowspan=0, so I'll implement that in a followup.

File: tests/wpt/mozilla/tests/css/table_rowspan_notequal_a.html [hunk 3]
```diff
@@ -0,0 +1,18 @@
+<!DOCTYPE html>
+<html>
+  <meta charset="UTF-8">
+  <link rel="match" href="table_rowspan_simple_ref.html">
```
Comment (jdm) [inline tests/wpt/mozilla/tests/css/table_rowspan_notequal_a.html]: Did you mean to use the new reference file that was added in this PR instead?

**Comment by jdm** [comment]
@bors-servo r-
Waiting for an answer.

**Comment by Manishearth** [comment]
Gah. Yep, fixed.

**Comment by jdm** [comment]
@bors-servo r-
The manifest needs to be updated.

**Pull request merged**